gneiss/util.py

```diff
@@ -0,0 +1,117 @@
+def match(x, y, intersect=False):
+    """ Sorts samples in metadata and contingency table in the same order.
+
+    Parameters
+    ----------
+    x : pd.DataFrame
+        Contingency table where samples correspond to rows and
+        features correspond to columns.
+    y: pd.DataFrame
+        Metadata table where samples correspond to rows and
+        explanatory metadata variables correspond to columns.
+    intersect : bool, optional
+        Specifies if only the intersection of samples in the
+        contingency table and the metadata table will returned.
+
+    Returns
+    -------
+    _x : pd.DataFrame
+        Filtered dataframe
+    _y : pd.DataFrame
+        Filtered dataframe
+    """
+    _x = x.sort_index()
+    _y = y.sort_index()
+    if intersect:
+        idx = set(_x.index) & set(_y.index)
+        idx = sorted(idx)
+        return _x.loc[idx], _y.loc[idx]
+    else:
+        if len(_x.index) != len(_y.index):
+            raise ValueError("`x` and `y` have incompatible sizes, "
+                             "`x` has %d rows, `y` has %d rows.  "
+                             "Consider setting `intersect=True`." %
+                             (len(_x.index), len(_y.index)))
+        return _x, _y
+
+
+def match_tips(table, tree, intersect=False):
+    """ Returns the OTU table and tree with matched tips.
+
+    Sorts the columns of the OTU table to match the tips in
+    the tree.  If the tree is multi-furcating, then the
+    tree is reduced to a bifurcating tree by randomly inserting
+    internal nodes.
+
+
+    Parameters
+    ----------
+    table : pd.DataFrame
+        Contingency table where samples correspond to rows and
+        features correspond to columns.
+    tree : skbio.TreeNode
+        Tree object where the leafs correspond to the features.
+    intersect : bool, optional
+        Specifies if only the intersection of samples in the
+        contingency table and the tree will returned.
+
+    Returns
+    -------
+    pd.DataFrame :
+        Subset of the original contingency table with the common features.
+    skbio.TreeNode :
+        Sub-tree with the common features.
+    """
+    tips = [x.name for x in tree.tips()]
+    common_tips = list(set(tips) & set(table.columns))
```

Comment:
Because you cannot use sets to index DataFrames

```python
In [1]: import pandas as pd
In [2]: table = pd.DataFrame([[0, 0, 1, 1],
   ...:                       [2, 2, 4, 4],
   ...:                       [5, 5, 3, 3],
   ...:                       [0, 0, 0, 1]],
   ...:                       index=['s1', 's2', 's3', 's4'],
   ...:                       columns=['o1', 'o2', 'o3', 'o4'])

In [3]: x = set(['o1','o2'])

In [4: table[x]
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-313a5d343876> in <module>()
----> 1 table[x]

/Users/mortonjt/miniconda3/envs/gneiss/lib/python3.5/site-packages/pandas/core/frame.py in __getitem__(self, key)
   1995             return self._getitem_multilevel(key)
   1996         else:
-> 1997             return self._getitem_column(key)
   1998 
   1999     def _getitem_column(self, key):

/Users/mortonjt/miniconda3/envs/gneiss/lib/python3.5/site-packages/pandas/core/frame.py in _getitem_column(self, key)
   2002         # get column
   2003         if self.columns.is_unique:
-> 2004             return self._get_item_cache(key)
   2005 
   2006         # duplicate columns & possible reduce dimensionality

/Users/mortonjt/miniconda3/envs/gneiss/lib/python3.5/site-packages/pandas/core/generic.py in _get_item_cache(self, item)
   1346         """Return the cached item, item represents a label indexer."""
   1347         cache = self._item_cache
-> 1348         res = cache.get(item)
   1349         if res is None:
   1350             values = self._data.get(item)

TypeError: unhashable type: 'set'
```